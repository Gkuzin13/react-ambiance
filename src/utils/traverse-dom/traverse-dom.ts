import { Children, createElement, isValidElement } from 'react';
import type { ReactNode, MutableRefObject, DOMAttributes } from 'react';
import type { SourceRef } from '@/components/AmbientCanvas/AmbientCanvas';

export type MediaType = 'video' | 'img';

export type MediaProps = {
  ref: MutableRefObject<any>;
} & DOMAttributes<SourceRef['current']>;

export const traverseAndPassPropsByElementType = (
  children: ReactNode,
  type: MediaType,
  props: MediaProps,
): ReactNode => {
  return Children.map(children, (child) => {
    if (!isValidElement(child)) return child;

    if (child.type === type) {
      return createElement(type, { ...props, ...child.props });
    } else if (child.props.children) {
      return traverseAndPassPropsByElementType(
        child.props.children,
        type,
        props,
      );
    }

    return child;
  });
};
