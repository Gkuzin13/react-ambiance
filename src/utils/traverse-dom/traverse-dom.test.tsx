import { test, vi } from 'vitest';
import { traverseAndPassPropsByElementType } from './traverse-dom';
import type { PropsWithChildren } from 'react';

test('passes props to element', () => {
  const propsToPass = {
    title: 'Test Title',
    onLoad: vi.fn(),
  };

  const Children = () => {
    return (
      <div>
        <img src="test" alt="test" />
      </div>
    );
  };

  const Wrapper = ({ children }: PropsWithChildren) => {
    return (
      <div>
        {traverseAndPassPropsByElementType(children, 'img', propsToPass)}
      </div>
    );
  };

  const childrenWithPassedProps = Wrapper({ children: Children() });

  const [title, onLoad] = Object.keys(propsToPass);

  expect(childrenWithPassedProps.props.children[0].props).toHaveProperty(
    title,
    propsToPass.title,
  );

  expect(childrenWithPassedProps.props.children[0].props).toHaveProperty(
    onLoad,
    propsToPass.onLoad,
  );

  expect(childrenWithPassedProps).toMatchSnapshot();
});
