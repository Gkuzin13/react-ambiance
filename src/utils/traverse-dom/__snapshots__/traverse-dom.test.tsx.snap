// Vitest Snapshot v1

exports[`passes props to element 1`] = `
<div>
  <img
    alt="test"
    onLoad={[MockFunction spy]}
    src="test"
    title="Test Title"
  />
</div>
`;
